### PR TITLE
Add support for custom coordinate frames in WCSAxes specifically in transform_coord_meta_from_wcs

### LINKED ISSUE #16339: Support custom mapping from UCD to WSCAxes properties for custom coordinate frames
### What is the problem this feature will solve?

The overall aim is to have nice plot with proper `ax.coords` setting e.g. `coord_wrap`, `format_unit`, `coord_type` for custom Coordinate frames.

Currently half of the is supported, `fitswcs.py` supports custom ctype to `UCD` mappings via `CTYPE_TO_UCD1_CUSTOM` but then on the other side `wscapi.py` has a hard coded list in `transform_coord_meta_from_wcs` therefore `WCSAxes` doesn't recognise the custom UCD.

So currently this can be done
```
MYFRAME_CTYPE_TO_UCD1 = {
    "SXLT": "custom:pos.myframe.lat",
    "SXLN": "custom:pos.myframe.lon",
    "SXRZ": "custom:pos.myframe.z"
}
astropy.wcs.wcsapi.fitswcs.CTYPE_TO_UCD1_CUSTOM.update(MYFRAME_CTYPE_TO_UCD1)
```
but there is no way to include settings for `custom:pos.myframe.X` in this section of code https://github.com/astropy/astropy/blob/cf99b9e7761c45c15ddc060a76b563574f525df9/astropy/visualization/wcsaxes/wcsapi.py#L54-L71

where Id like to add something like
```
            if "custom:pos.MYFRAME.lon" in axis_type:
                coord_wrap = 180.0 * u.deg
                format_unit = u.arcsec
                coord_type = "longitude"
            elif "custom:pos.MYFARME.lat" in axis_type:
                format_unit = u.arcsec
                coord_type = "latitude"
```

### Describe the desired outcome

Be able to register or other wise include custom plot setting for custom UCDs in a similar manner to how custom CTYPE to UCD mapping can be included.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by samaloney**
### Description

This pull request adds support for externally created customs coordinate frames to WCSAxes by altering how `transform_coord_meta_from_wcs` operates.
-  Creating a new variable `CUSTOM_UCD_COORD_META_MAPPING` which could be extended by external packages.
-  Move existing custom frame to this new variable

If this approach is ok then the custom sunpy frame could be moved into the sunpy repo.

~~I didn't add a new test as while there is new code the current tests cover it I think.~~

In an external package I could register new CTYPE to UCD and  UCD to coord meta data mappings.

```
MYFRAME_CTYPE_TO_UCD1 = {
    "SXLT": "custom:pos.myframe.lat",
    "SXLN": "custom:pos.myframe.lon",
    "SXRZ": "custom:pos.myframe.z"
}
astropy.wcs.wcsapi.fitswcs.CTYPE_TO_UCD1_CUSTOM.update(MYFRAME_CTYPE_TO_UCD1)


MYFAME_UCD_COORD_META_MAPPING = {
    "custom:pos.myframe.lon": {
        "coord_wrap": 180.0 * u.deg,
        "format_unit": u.arcsec,
        "coord_type": "longitude",
    },

astropy.visualization.wcsaxes.wcsapi.CUSTOM_UCD_COORD_META_MAPPING.update(MYFAME_UCD_COORD_META_MAPPING)
```

Fixes #16339 

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/visualization/wcsaxes/wcsapi.py
Comment (neutrinoceros) [inline astropy/visualization/wcsaxes/wcsapi.py:107]: I think this is needed to preserve existing behavior (unless the change is intentional ?)
```suggestion
                if ucd in axis_type:
                    dim_meta.update(meta)
                    break
```

**Review by neutrinoceros — COMMENTED** [review]
Thank you for your contribution ! I'll leave it to wcsaxes maintainers to discuss whether this approach is acceptable or not, but here's a quick technical review.
Even though the code might be covered already, I believe a test would also be needed to demonstrate the new feature (and to make sure we don't break it in the future).
You'll also need to provide a change log entry (see `docs/changes/README.md`). Thanks !

File: astropy/visualization/wcsaxes/tests/test_wcsapi.py
Comment (astrofrog) [inline astropy/visualization/wcsaxes/tests/test_wcsapi.py:281]: I think we should avoid having this be the API to register custom mappings, and instead write a context manager similar to ``custom_ctype_to_ucd_mapping`` to make sure the changes are temporary and cleaned up after use.

Comment (astrofrog) [reply astropy/visualization/wcsaxes/tests/test_wcsapi.py:281]: This would also have the benefit of allowing this functionality to be documented via the API docs even if we don't add something to the main narrative docs.

**Review by astrofrog — CHANGES_REQUESTED** [review]
This is looking good! Just one comment below.

File: astropy/visualization/wcsaxes/tests/test_wcsapi.py
Comment (samaloney) [reply astropy/visualization/wcsaxes/tests/test_wcsapi.py:282]: Great thanks for the feedback, I've taken a first crack at adding a context manager similar `custom_ctype_to_ucd_mapping`.

Comment (astrofrog) [reply astropy/visualization/wcsaxes/tests/test_wcsapi.py:282]: Thanks! I don't think we should keep the test of modifying the dictionary, we should just be testing the context manager.

File: astropy/visualization/wcsaxes/wcsapi.py
Comment (astrofrog) [inline astropy/visualization/wcsaxes/wcsapi.py:104]: Ideally we might want to allow existing entries to be overriden, but if this is too much work we could always keep as-is for now and consider it a future improvement to allow overriding.

File: astropy/visualization/wcsaxes/tests/test_wcsapi.py
Comment (astrofrog) [inline astropy/visualization/wcsaxes/tests/test_wcsapi.py:330]: It would be good if possible to add a test of two nested ``custom_ucd_wcscoord_mapping`` calls to make sure both apply at the same time.

**Review by astrofrog — CHANGES_REQUESTED** [review]
Looking good! Just a few small comments.

File: astropy/visualization/wcsaxes/tests/test_wcsapi.py
Comment (Cadair) [reply astropy/visualization/wcsaxes/tests/test_wcsapi.py:282]: Given there's no way to stop someone editing the dictionary shouldn't we keep a test for it?

File: astropy/visualization/wcsaxes/tests/test_wcsapi.py
Comment (Cadair) [reply astropy/visualization/wcsaxes/tests/test_wcsapi.py:330]: Added

File: astropy/visualization/wcsaxes/wcsapi.py
Comment (Cadair) [reply astropy/visualization/wcsaxes/wcsapi.py]: Patched in opt-in support for overwriting (with a kwarg)

File: astropy/visualization/wcsaxes/tests/test_wcsapi.py
Comment (neutrinoceros) [reply astropy/visualization/wcsaxes/tests/test_wcsapi.py:282]: I'd vote against it; we shouldn't be testing for implementation details.

File: astropy/visualization/wcsaxes/wcsapi.py
Comment (neutrinoceros) [reply astropy/visualization/wcsaxes/wcsapi.py]: oh my. I didn't realize this comment never got public. Sorry about that.

File: astropy/visualization/wcsaxes/tests/test_wcsapi.py
Comment (astrofrog) [reply astropy/visualization/wcsaxes/tests/test_wcsapi.py:282]: If the dictionary isn't meant to be accessed directly maybe we should make it private to make it explicit?

File: astropy/visualization/wcsaxes/wcsapi.py
Comment (Cadair) [reply astropy/visualization/wcsaxes/wcsapi.py]: Fixed thanks

File: astropy/visualization/wcsaxes/tests/test_wcsapi.py
Comment (Cadair) [reply astropy/visualization/wcsaxes/tests/test_wcsapi.py]: I removed it, I don't think we need to make it private

Comment (Cadair) [reply astropy/visualization/wcsaxes/tests/test_wcsapi.py]: I do think there is utility in people directly accessing the dict if packages like sunpy write their own coordinate frames they could add to the dict without the context manager to improve the plotting of those frames.

File: docs/changes/visualization/16347.feature.rst
Comment (astrofrog) [inline docs/changes/visualization/16347.feature.rst:1]: Should we maybe not mention the dictionary here since it's an implementation detail?

**Review by astrofrog — CHANGES_REQUESTED** [review]
This seems good overall but just a small note on the changelog entry - and in addition I though ``wcscoord`` in the context manager name was a bit strange, what about using ``custom_ucd_coord_meta_mapping`` as the context manager name?

**Comment by Cadair** [comment]
Lgtm thanks

**Review by astrofrog — APPROVED** [review]

**Pull request merged**
